synth/src/datasource/postgres_datasource.rs

```diff
@@ -0,0 +1,330 @@
+use crate::datasource::DataSource;
+use anyhow::{Result, Context};
+use crate::datasource::relational_datasource::{RelationalDataSource, ColumnInfo, PrimaryKey, ForeignKey, ValueWrapper};
+use sqlx::{Pool, Postgres, Row, Column, TypeInfo};
+use sqlx::postgres::{PgPoolOptions, PgQueryResult, PgRow, PgColumn};
+use serde_json::{Value, Map, Number};
+use async_std::task;
+use async_trait::async_trait;
+use std::convert::TryFrom;
+use rust_decimal::Decimal;
+use rust_decimal::prelude::ToPrimitive;
+use synth_core::Content;
+use synth_core::schema::{BoolContent, StringContent, RegexContent, NumberContent, RangeStep, DateTimeContent, ChronoValueType, Uuid};
+use synth_core::schema::number_content::{I64, F64};
+
+pub struct PostgresDataSource {
+    pool: Pool<Postgres>,
+    single_thread_pool: Pool<Postgres>,
+    connect_params: String
+}
+
+#[async_trait]
+impl DataSource for PostgresDataSource {
+    type ConnectParams = String;
+
+    fn new(connect_params: &Self::ConnectParams) -> Result<Self> {
+        task::block_on(async {
+            let pool = PgPoolOptions::new()
+                .max_connections(3) //TODO expose this as a user config?
+                .connect(connect_params.as_str())
+                .await?;
+
+            // Needed for queries that require explicit synchronous order, i.e. setseed + random
+            let single_thread_pool = PgPoolOptions::new()
+                .max_connections(1)
+                .connect(connect_params.as_str())
+                .await?;
+
+            Ok::<Self, anyhow::Error>(PostgresDataSource {
+                pool,
+                single_thread_pool,
+                connect_params: connect_params.to_string()
+            })
+        })
+    }
+
+    async fn insert_data(&self, collection_name: String, collection: &[Value]) -> Result<()> {
+        self.insert_relational_data(collection_name, collection).await.unwrap();
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl RelationalDataSource for PostgresDataSource {
+    type QueryResult = PgQueryResult;
+
+    async fn execute_query(&self, query: String, _query_params: Vec<&str>) -> Result<PgQueryResult> {
+        let result = sqlx::query(query.as_str())
+            .execute(&self.pool)
+            .await?;
+
+        Ok(result)
+    }
+
+    fn get_catalog(&self) -> Result<&str> {
```

Comment:
This is how we got the catalog in the original implementation as well. I have since found a better way (and I think it's possible in SQLX as well.

You can use the `ConnectionOptions` which will parse the URI and get the `dbname` from there. See: https://docs.rs/sqlx/0.5.5/sqlx/postgres/struct.PgConnectOptions.html